datalad_next/runners/data_processors.py

```diff
@@ -0,0 +1,316 @@
+"""
+This module contains a generator that reads data-chunks from an iterable,
+passes it through a number of data processors and yields the result of the
+last data processor
+
+Typical data processors would be:
+
+- decode a stream of bytes
+- split a stream of characters at line-ends
+- convert a line of text into JSON
+
+Data processors have a common interface and can be chained. For example,
+one can pass data chunks, where each chunk is a byte-string, into a chain
+of two data processors: a decode-processor that converts bytes into strings,
+and a linesplit-processor that converts character-streams into lines. The result
+of the chain would be lines of text.
+
+Data processors might need to buffer some data before yielding their result. The
+"driver" of the data processing chains supports the buffering of input data for
+individual processors. Therefore, most data processors do not need to store
+state themselves and can be quite simple
+"""
+from __future__ import annotations
+
+import json
+from collections import defaultdict
+from collections.abc import Generator
+from functools import partial
+from typing import (
+    Any,
+    Callable,
+    Iterable,
+    List,
+    Union,
+)
+
+from datalad_next.runners import LineSplitter
+
+
+# Protocol machines are callables that have the following signature:
+#
+#    def process(self, data: list[T], final: bool) -> tuple[list[N] | None, list[T]]:
+#       ...
+#
+# where N is the type that is returned by processor. The return value is a
+# consisting of optional results, i.e. list[N] | None, and a number of input
+# elements that were not processed and should be presented again, when more
+# data arrives from the "preceding" element.
+
+StrList = List[str]
+BytesList = List[bytes]
+StrOrBytes = Union[str, bytes]
+StrOrBytesList = List[StrOrBytes]
+
+
+def jsonline_processor(lines: StrOrBytesList,
+                       _: bool = False
+                       ) -> tuple[list[tuple[bool, Any]], StrOrBytesList]:
```

Comment:
Seems tested, but is otherwise unused.

Reply:
Yes, it is unused. It could be used to process JSON output from `git annex --json ...` calls.